src/natcap/invest/coastal_blue_carbon/coastal_blue_carbon.py

```diff
@@ -188,941 +194,1521 @@
-            d[args_key] = dict(
-                (code, row[col_name]) for code, row in code_dict.items())
-
-    # Transition Dictionaries
-    biomass_transient_dict, soil_transient_dict = \
-        _create_transient_dict(args['carbon_pool_transient_uri'])
-
-    d['lulc_to_Yb'] = dict((key, sub['yearly-accumulation'])
-                           for key, sub in biomass_transient_dict.items())
-    d['lulc_to_Ys'] = dict((key, sub['yearly-accumulation'])
-                           for key, sub in soil_transient_dict.items())
-    d['lulc_to_Hb'] = dict((key, sub['half-life'])
-                           for key, sub in biomass_transient_dict.items())
-    d['lulc_to_Hs'] = dict((key, sub['half-life'])
-                           for key, sub in soil_transient_dict.items())
-
-    # Parse LULC Transition CSV (Carbon Direction and Relative Magnitude)
-    d['lulc_trans_to_Db'], d['lulc_trans_to_Ds'] = _get_lulc_trans_to_D_dicts(
-        args['lulc_transition_matrix_uri'],
-        args['lulc_lookup_uri'],
-        biomass_transient_dict,
-        soil_transient_dict)
-
-    # Economic Analysis
-    d['do_economic_analysis'] = False
-    if args['do_economic_analysis']:
-        d['do_economic_analysis'] = True
-        # convert percentage to decimal
-        discount_rate = float(args['discount_rate']) * 0.01
-        if args['do_price_table']:
-            d['price_t'] = _get_price_table(
-                args['price_table_uri'],
-                d['snapshot_years'][0],
-                d['snapshot_years'][-1])
-        else:
-            inflation_rate = float(args['inflation_rate']) * 0.01
-            price = float(args['price'])
-            d['price_t'] = (1 + inflation_rate) ** numpy.arange(
-                0, float(d['timesteps'])+1) * price
+    Returns:
+        ``None``.
 
-        d['price_t'] /= (1 + discount_rate) ** numpy.arange(0, d['timesteps']+1)
+    """
+    LOGGER.info('Summing %s rasters to %s', len(raster_path_list),
+                target_raster_path)
+    pygeoprocessing.new_raster_from_base(
+        raster_path_list[0], target_raster_path, gdal.GDT_Float32,
+        [NODATA_FLOAT32])
+
+    target_raster = gdal.OpenEx(
+        target_raster_path, gdal.GA_Update | gdal.OF_RASTER)
+    target_band = target_raster.GetRasterBand(1)
+    for block_info in pygeoprocessing.iterblocks(
+            (raster_path_list[0], 1), offset_only=True):
+
+        sum_array = numpy.empty(
+            (block_info['win_ysize'], block_info['win_xsize']),
+            dtype=numpy.float32)
+        sum_array[:] = 0.0
+
+        # Assume everything is valid until proven otherwise
+        valid_pixels = numpy.ones(sum_array.shape, dtype=numpy.bool)
+        pixels_touched = numpy.zeros(sum_array.shape, dtype=numpy.bool)
+        for raster_path in raster_path_list:
+            raster = gdal.OpenEx(raster_path, gdal.OF_RASTER)
+            band = raster.GetRasterBand(1)
+            band_nodata = band.GetNoDataValue()
```

Comment:
I imagine you've thought this through but I'm a little surprised to see you'd want to open and close file on every block iteration. Is there a possibility there might be too many files open at once? If so I think this is fine rather than some kind of dance to try to keep track of the pixels_touched array. Otherwise would you consider a list of raster/band/nodata tuples to avoid opening and closing N files for M iterblocks?

Reply:
Possibly?  The hope with opening/closing each raster was to hopefully avoid issues of too many open files.  I've implemented the update you suggested in 82d7cb61, which also includes some bonus timed logging, since this function can take a while to run on enough sufficiently-large rasters.

Reply:
There's more discussion related to this in https://github.com/natcap/invest/pull/316#discussion_r501258814 (comment below), and I've re-patched this in f118a2b2 to read in a block at a time since opening GDAL rasters seems to have a negligible effect on the overall runtime of the program.